test/invariants/fuzz/Protocol.t.sol

```diff
@@ -0,0 +1,568 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.25;
+
+import {EchidnaTest, FuzzERC20} from '../helpers/AdvancedTestsUtils.sol';
+
+import {MockSettler} from '../helpers/MockSettler.sol';
+import {BCoWFactory, BCoWPool, IBPool} from 'contracts/BCoWFactory.sol';
+import {BConst} from 'contracts/BConst.sol';
+import {BMath} from 'contracts/BMath.sol';
+import {BNum} from 'contracts/BNum.sol';
+
+contract FuzzProtocol is EchidnaTest {
+  // System under test
+  BCoWFactory factory;
+  BConst bconst;
+  BMath bmath;
+  BNum_exposed bnum;
+
+  address solutionSettler;
+  bytes32 appData;
+
+  FuzzERC20[] tokens;
+  BCoWPool pool;
+
+  IBPool[] poolsToFinalize;
+
+  uint256 ghost_bptMinted;
+  uint256 ghost_bptBurned;
+  mapping(FuzzERC20 => uint256) ghost_amountDirectlyTransfered;
+
+  constructor() {
+    solutionSettler = address(new MockSettler());
+
+    factory = new BCoWFactory(solutionSettler, appData);
+    bconst = new BConst();
+    bmath = new BMath();
+    bnum = new BNum_exposed();
+
+    // max bound token is 8
+    for (uint256 i; i < 4; i++) {
+      FuzzERC20 _token = new FuzzERC20();
+      _token.initialize('', '', 18);
+      tokens.push(_token);
+    }
+
+    pool = BCoWPool(address(factory.newBPool()));
+
+    for (uint256 i; i < 4; i++) {
+      FuzzERC20 _token = new FuzzERC20();
+      _token.initialize('', '', 18);
+      tokens.push(_token);
+
+      _token.mint(address(this), 10 ether);
+      _token.approve(address(pool), 10 ether);
+
+      uint256 _poolWeight = bconst.MAX_WEIGHT() / 5;
+
+      try pool.bind(address(_token), 10 ether, _poolWeight) {}
+      catch {
+        assert(false);
+      }
+    }
+
+    pool.finalize();
+    ghost_bptMinted = pool.INIT_POOL_SUPPLY();
+  }
+
+  // Randomly add or remove tokens to a pool
+  // Insure caller has enough token
+  // Main objective is to have an arbitrary number of tokens in the pool, peripheral objective is another
+  // test of min/max token bound (properties 20 and 21)
+  function setup_joinExitPool(bool _join, uint256 _amountBpt) public AgentOrDeployer {
+    if (_join) {
+      uint256[] memory _maxAmountsIn;
+
+      _maxAmountsIn = new uint256[](4);
+
+      for (uint256 i; i < _maxAmountsIn.length; i++) {
+        uint256 _maxIn =
+          bnum.bmul_exposed(bnum.bdiv_exposed(_amountBpt, pool.totalSupply()), pool.getBalance(address(tokens[i])));
+        _maxAmountsIn[i] = _maxIn;
+
+        tokens[i].mint(currentCaller, _maxIn);
+        hevm.prank(currentCaller);
+        tokens[i].approve(address(pool), _maxIn);
+      }
+
+      hevm.prank(currentCaller);
+      try pool.joinPool(_amountBpt, _maxAmountsIn) {
+        ghost_bptMinted += _amountBpt;
+      } catch {
+        assert(
+          pool.isFinalized() || pool.getCurrentTokens().length > bconst.MAX_BOUND_TOKENS()
+            || currentCaller != pool.getController()
+        );
+      }
```

Comment:
after a run of 50k calls I'm getting no coverage on this, perhaps there's something wrong with line 80?

the calls to `_{mint,push}PoolShare` in `join` are also not covered